src/main/java/com/aws/greengrass/lifecyclemanager/GenericExternalService.java

```diff
@@ -238,13 +238,45 @@ public boolean isBootstrapRequired(Map<String, Object> newServiceConfig) {
         }
         Node serviceOldBootstrap = getConfig().findNode(SERVICE_LIFECYCLE_NAMESPACE_TOPIC,
                 Lifecycle.LIFECYCLE_BOOTSTRAP_NAMESPACE_TOPIC);
-        boolean bootstrapStepChanged =  serviceOldBootstrap == null
-                || !serviceOldBootstrap.toPOJO().equals(newServiceLifecycle.get(lifecycleKey));
-        logger.atDebug().log(String.format("Bootstrap is %srequired: bootstrap step %schanged",
-                bootstrapStepChanged ? "" : "not ", bootstrapStepChanged ? "" : "un"));
+        boolean bootstrapStepChanged = serviceOldBootstrap == null || !serviceLifecycleBootstrapEquals(
+                serviceOldBootstrap.toPOJO(), newServiceLifecycle.get(lifecycleKey));
+        if (bootstrapStepChanged) {
+            logger.atDebug().kv("before", serviceOldBootstrap.toPOJO())
+                    .kv("after", newServiceLifecycle.get(lifecycleKey))
+                    .log("Bootstrap is required: bootstrap step changed");
+        } else {
+            logger.atDebug().log("Bootstrap is not required: bootstrap step unchanged");
+        }
         return bootstrapStepChanged;
     }
 
+    private boolean serviceLifecycleBootstrapEquals(Object before, Object after) {
+        if (!(before instanceof Map) || !(after instanceof Map)) {
```

Comment:
should be `&&`, right?

Reply:
What I meant is, we also need to check the situation when one is map, and the other is not a map.

Reply:
When one config value is map, and the other is not a map, they definitely are not equal and I think it should be covered with `Objects.equal`